Improve the dependency path exception message

andriy-dmytruk opened the pull request:
This improves the readability of message for long paths, large class names or classes with many parameters, as each class from the path is immediately visible. I changed the tests to use longer class names, which is more realistic.

The type names are also in fully-qualified shortened form, which makes them clickable in IDEs.

As a comparison this message
```java
Message: No bean of type [io.micronaut.inject.failures.NestedDependencyFailureSpec$MyClassD] exists. 
| Path Taken: new MyClassB() --> MyClassB.propA --> new MyClassA([MyClassC propC]) --> new MyClassC([MyClassD propD])
```
will change to
```java
Message: No bean of type [io.micronaut.inject.failures.NestedDependencyFailureSpec$MyClassD] exists.
Path Taken:
new i.m.i.f.N$MyClassB()
\---> i.m.i.f.N$MyClassB#propA
      \---> new i.m.i.f.N$MyClassA([MyClassC propC])
            \---> new i.m.i.f.N$MyClassC([MyClassD propD])
```

Part of the change is similar to visualization change for circular dependencies: https://github.com/micronaut-projects/micronaut-core/pull/11299.

andriy-dmytruk committed:
Improve the dependency path exception message by printing each segment on a new line.

This improves the readability of message for long paths, large classnames or classes with many parameters, as each class from the path is immediately visible. I changed the tests to use longer classnames, which is more realistic.

andriy-dmytruk committed:
Remove duplication

sdelamo commented:
maybe this should target 4.8.x

dstepanov commented:
Consider adding a full class name. I think IntelliJ will provide a link to the class

dstepanov commented:
How does it look with a factory method / field?

andriy-dmytruk commented on inject-groovy/src/test/groovy/io/micronaut/inject/failures/NestedDependencyFailureSpec.groovy:
```diff
@@ -28,44 +28,49 @@ class NestedDependencyFailureSpec extends Specification {
+Message: No bean of type [io.micronaut.inject.failures.NestedDependencyFailureSpec\$MyClassD] exists.$space
+Path Taken:$space
+new MyClassB()
+\\---> MyClassB.propA
```
There is an example of field here

andriy-dmytruk commented:
> How does it look with a factory method / field?

For factory there is a circular dependency test: https://github.com/micronaut-projects/micronaut-core/blob/da67f65f446a2c75c6e1048139e88d8650332f2b/inject-groovy/src/test/groovy/io/micronaut/inject/failures/FactoryCircularDependencyFailureSpec.groovy#L41-L48

graemerocher commented:
> Consider adding a full class name. I think IntelliJ will provide a link to the class

this is a good point, with the new lines maybe we can include qualified names. Can you check this link works in IntelliJ?

andriy-dmytruk commented:
Yes, the classnames are clickable:
<img width="1113" alt="image" src="https://github.com/user-attachments/assets/f443b36a-0250-42d1-9467-7185caba0986">

This is more readable, though:
<img width="1143" alt="image" src="https://github.com/user-attachments/assets/73393e9b-2437-4c7b-8d74-c7fe87082489">

I think both are good. Wdyt?

graemerocher commented:
@dstepanov WDYT?

dstepanov commented:
Intellij supports even shortened names, so we can reduce it if it's too large:

try:
```
    public static void main(String[] args) {
        System.out.println("io.m.r.Micronaut!");
    }
```

graemerocher commented:
@andriy-dmytruk can you try that?

andriy-dmytruk commented:
Yes, this is actually amazing! Now there are actually 2 options:
1. Shorten type owner names: 
    <img width="1087" alt="image" src="https://github.com/user-attachments/assets/e7353cae-0287-42de-abfb-f3163413f30a">
  When clicked:
    <img width="1072" alt="image" src="https://github.com/user-attachments/assets/257e8403-69a6-4f4d-bb9f-df21977993e1">

2. Do not shorten:
    <img width="1075" alt="image" src="https://github.com/user-attachments/assets/6a582178-6923-470b-b30b-ab535703238e">

I would say the first one is the overall best.

dstepanov commented:
Maybe we can have something like 200-300 characters per name and start reducing when it's more.

graemerocher commented:
the first one is the best yes

andriy-dmytruk committed:
Use shortened fully-qualified names in dependency injection errors

Fully-qualified names are clickable in IDE terminals, which improves usability. Shortened form prevents the messages from being too convoluted.

andriy-dmytruk commented:
I went for using the first form always, as it is shortest, but is still clickable. We can also extend it at a later point if we experience issues with it.

> Maybe we can have something like 200-300 characters per name and start reducing when it's more.

We can start a GitHub discussion, but I think shortest should be good in most cases. The chances of conflict are pretty low. If we do need this, the typical line limits are 80, 100, 120, which would mean around 60 chars limit for the name alone.

graemerocher commented on core/src/main/java/io/micronaut/core/naming/NameUtils.java:
```diff
@@ -214,6 +215,41 @@ public static String getSimpleName(String className) {
+     * @return The shortened type name
+     */
+    @Experimental
+    public static String getShortenedName(String typeName) {
```
add nullability annotations to return type and arguments

graemerocher commented on core/src/main/java/io/micronaut/core/naming/NameUtils.java:
```diff
@@ -214,6 +215,41 @@ public static String getSimpleName(String className) {
+     */
+    @Experimental
+    public static String getShortenedName(String typeName) {
+        int nameStart = typeName.lastIndexOf('$');
```
check type name doesn't == null

andriy-dmytruk replied:
That is not checked in other methods, so I assume it is not typical for `typeName` to be `null`. Added `@NonNull` though.

graemerocher reviewed (approved):
just two small tweaks before merging.

andriy-dmytruk committed:
Add nullability annotations

andriy-dmytruk commented:
@yawkat I don't understand the issue. Can you take a look?

https://ge.micronaut.io/s/szoo56vrxva66
<img width="872" alt="44C5523F-E220-4E09-B7D3-AB5A84F2010C" src="https://github.com/user-attachments/assets/4e48975b-c138-4f57-a50f-851a1b14443b">

yawkat commented:
that test is just flakey. i dont know why
